### PR TITLE
Add arm64e slice

### PR DESCRIPTION
**Pull request by ElektrojungeMS**
Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* ~[ ] Did you add unit tests?~
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

This is sort of an FYI PR to add visibility for the arm64e work we are doing.

### CODE REVIEW

File: CrashLib/CrashLib.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -934,6 +934,11 @@
 		B21A71E41DD2A38C0044BB1C /* Debug */ = {
 			isa = XCBuildConfiguration;
 			buildSettings = {
+				ARCHS = (
```
Comment (jaelim-ms) [inline CrashLib/CrashLib.xcodeproj/project.pbxproj]: Can we add a new xcconfig for `CrashLib`?

File: Sasquatch/Config/Sasquatch.xcconfig [hunk 2]
```diff
@@ -1,5 +1,12 @@
 #include "../../Config/App.xcconfig"
 
+ARCHS[sdk=iphoneos*] = armv7 armv7s arm64 arm64e
+ARCHS[sdk=watchos*] = armv7k arm64_32
+ARCHS[sdk=iphonesimulator*] = x86_64 i386
```
Comment (jaelim-ms) [inline Sasquatch/Config/Sasquatch.xcconfig]: We have `MS_SIM_ARCHS` defined in `Global.xcconfig`.

File: Sasquatch/Config/Sasquatch.xcconfig [hunk 3]
```diff
@@ -1,5 +1,12 @@
 #include "../../Config/App.xcconfig"
 
+ARCHS[sdk=iphoneos*] = armv7 armv7s arm64 arm64e
+ARCHS[sdk=watchos*] = armv7k arm64_32
```
Comment (jaelim-ms) [inline Sasquatch/Config/Sasquatch.xcconfig]: We have `MS_WATCH_ARCHS` defined in `Global.xcconfig`. Let's add `arm64_32` to global config and use it.

**Comment by ElektrojungeMS** [comment]
Keeping this as open for verification purposes. We'll clean things up once we are done verifying things and will update this once it's ready for review and merging.

**Comment by ElektrojungeMS** [comment]
f25a94a3355190d62404a3f3b3e30ca745f3f854 adds a PLCR build that won't work in Xcode 8. This is just for verification purposes.

File: CrashLib/CrashLib.xcodeproj/project.pbxproj (hunk 1 above)
Comment (ElektrojungeMS) [reply CrashLib/CrashLib.xcodeproj/project.pbxproj]: I'll create a separate bug for this.

File: Sasquatch/Config/Sasquatch.xcconfig (hunk 3 above)
Comment (ElektrojungeMS) [reply Sasquatch/Config/Sasquatch.xcconfig]: I've added that.

File: Sasquatch/Config/Sasquatch.xcconfig (hunk 2 above)
Comment (ElektrojungeMS) [reply Sasquatch/Config/Sasquatch.xcconfig]: the apps are currently NOT using `Global.xcconfig` and I'd rather not introduce this change in this PR.

File: CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,10 @@
 * **[Fix]** Fix reporting carrier information using new iOS 12 APIs when running on iOS 12+.
 * **[Fix]** Fix a memory leak issue during executing SQL queries.
 
+### AppCenterCrashes
```
Comment (clpolet) [inline CHANGELOG.md]: The arm64e slice is there for all modules, maybe we should add lines to other modules as well?

File: Config/iOS.xcconfig [hunk 5]
```diff
@@ -1,8 +1,17 @@
...
+PRE_XCODE_10_1_ARCHS = armv7 armv7s arm64
+ARCHS[sdk=iphoneos*] = $(PRE_XCODE_10_1_ARCHS) arm64e
+ARCHS[sdk=iphoneos9.*] = $(PRE_XCODE_10_1_ARCHS)
+ARCHS[sdk=iphoneos10.*] = $(PRE_XCODE_10_1_ARCHS)
+ARCHS[sdk=iphoneos11.*] = $(PRE_XCODE_10_1_ARCHS)
+ARCHS[sdk=iphoneos12.0] = $(PRE_XCODE_10_1_ARCHS)
 ARCHS[sdk=iphonesimulator*] = $(MS_SIM_ARCHS)
-VALID_ARCHS = $(ARCHS)
+VALID_ARCHS[sdk=iphoneos*] = $(ARCHS)
+VALID_ARCHS[sdk=iphoneos9.*] = $(ARCHS)
```
Comment (jaelim-ms) [inline Config/iOS.xcconfig]: Do you need line 11-14?

File: Scripts/AppCenterAnalytics/build-appcenteranalytics-tvos-framework.sh [hunk 6]
```diff
@@ -41,7 +41,7 @@ cp -f "${SRCROOT}/${FMK_NAME}/Support/tvOS.modulemap" "${INSTALL_DIR}/Modules/mo
 # Copies the headers and resources files to the final product folder.
 cp -R "${SRCROOT}/${WRK_DIR}/Release-appletvos/include/${FMK_NAME}/" "${INSTALL_DIR}/Headers/"
 
-# Uses the Lipo Tool to merge both binary files (i386 + arm64) into one Universal final product.
+# Uses the Lipo Tool to merge both binary files (i386/x86_64 + armv7/armv7s/arm64) into one Universal final product.
```
Comment (jaelim-ms) [inline Scripts/AppCenterAnalytics/build-appcenteranalytics-tvos-framework.sh]: Is this a correct change?

File: Scripts/AppCenter/build-appcenter-ios-framework.sh [hunk 7]
```diff
@@ -1,5 +1,8 @@
 #!/bin/sh
 
+# Load local .bash_profile to have access to MS_ARM64E_XCODE_PATH
+source ~/.bash_profile
```
Comment (jaelim-ms) [inline Scripts/AppCenter/build-appcenter-ios-framework.sh]: Should we do this in the script? It is obvious that any of changes made in `profile`s wouldn't be applied right away.
When customers add environment variable to their `profile` files, they should run `source` to apply the changes.
It isn't a big deal to us but users might have some other changes in their `profile` files which are unrelated to App Center.

File: Scripts/AppCenter/build-appcenter-ios-framework.sh [hunk 8]
```diff
@@ -41,8 +44,32 @@ cp -f "${SRCROOT}/${FMK_NAME}/Support/iOS.modulemap" "${INSTALL_DIR}/Modules/mod
 # Copies the headers and resources files to the final product folder.
 cp -R "${SRCROOT}/${WRK_DIR}/Release-iphoneos/include/${FMK_NAME}/" "${INSTALL_DIR}/Headers/"
 
-# Uses the Lipo Tool to merge both binary files (i386 + armv6/armv7/armv7s/arm64) into one Universal final product.
-lipo -create "${DEVICE_DIR}/lib${FMK_NAME}.a" "${SIMULATOR_DIR}/lib${FMK_NAME}.a" -output "${INSTALL_DIR}/${FMK_NAME}"
+# Create the arm64e slice in Xcode 10.1 and lipo it with the device binary that was created with oldest supported Xcode version.
+# Move binary that was create with old Xcode to temp location.
+LIB_IPHONEOS_FINAL="${DEVICE_DIR}/lib${FMK_NAME}.a"
+if [ -z "$MS_ARM64E_XCODE_PATH" ] || [ ! -d "$MS_ARM64E_XCODE_PATH" ] ; then
```
Comment (jaelim-ms) [inline Scripts/AppCenter/build-appcenter-ios-framework.sh]: `[ -z "$MS_ARM64E_XCODE_PATH" ] || [ ! -d "$MS_ARM64E_XCODE_PATH" ]` condition is used twice in the file. You can declare a local variable or you can combine two `if` statements.
This script shows it like
```
if (condition) {
  // Print error message
} else {
  // Building arm64e slice
}
if (condition) {
  // Use lagacy Xcode to generate fat binary
} else {
  // Use arm64e Xcode to generate fat binary
}
```
You can do it like
```
if (condition) {
  // Print error message
  // Use lagacy Xcode to generate fat binary
} else {
  // Building arm64e slice
  // Use arm64e Xcode to generate fat binary
}
```

File: Scripts/AppCenter/build-appcenter-ios-framework.sh [hunk 9]
```diff
@@ -41,8 +44,32 @@ cp -f "${SRCROOT}/${FMK_NAME}/Support/iOS.modulemap" "${INSTALL_DIR}/Modules/mod
...
+echo "Environment variable MS_ARM64E_XCODE_PATH not set or not a valid path."
+else
+echo "Building the arm64e slice."
+LIB_IPHONEOS_TEMP_DIR="${DEVICE_DIR}/temp"
+mkdir -p "${LIB_IPHONEOS_TEMP_DIR}"
+mv "${DEVICE_DIR}/lib${FMK_NAME}.a" "${LIB_IPHONEOS_TEMP_DIR}/lib${FMK_NAME}.a"
+
+# Build with the Xcode version that supports arm64e.
+env DEVELOPER_DIR="${MS_ARM64E_XCODE_PATH}" /usr/bin/xcodebuild ARCHS="arm64e" -project "${FMK_NAME}.xcodeproj" -configuration "Release" -target "${TGT_NAME}"
```
Comment (jaelim-ms) [inline Scripts/AppCenter/build-appcenter-ios-framework.sh]: Are you using `DEVELOPER_DIR` variable? You keep assigning the result to the env variable.

File: README.md [hunk 10]
```diff
@@ -25,7 +25,12 @@ It is super easy to use App Center. Have a look at our [get started documentatio
 
 We are looking forward to your contributions via pull requests.
 
-To contribute to the SDK, you need a machine that runs macOS and the latest stable Xcode release. In addition, the SDK workspace requires [Jazzy](https://github.com/realm/jazzy) to generate documentation. 
+To contribute to the SDK, please
+
+* install Xcode 10.1 and Xcode 8.3.3 in parallel (information on how to do that can be found [in this blogpost](https://medium.com/@hacknicity/working-with-multiple-versions-of-xcode-e331c01aa6bc)).
```
Comment (jwallra) [inline README.md]: Since I ran into this problem, we should mention that 8.x can't be used on Mojave.  Or is it just the older MacBook Pros that it can't be used on?  What if someone can't install multiple versions for this reason, should we just say that they can't build the SDK for the correct architectures?  I guess I don't really know what path to development we want to give people in those situations

File: Sasquatch/Sasquatch/Base.lproj/LaunchScreen.storyboard [hunk 11]
```diff
@@ -1,11 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="13529" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" launchScreen="YES" useTraitCollections="YES" colorMatched="YES" initialViewController="D7o-BS-rls">
-    <device id="retina5_9" orientation="portrait">
+<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="12121" systemVersion="17G3025" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" launchScreen="YES" useTraitCollections="YES" colorMatched="YES" initialViewController="D7o-BS-rls">
+    <device id="retina4_7" orientation="portrait">
```
Comment (jwallra) [inline Sasquatch/Sasquatch/Base.lproj/LaunchScreen.storyboard]: Why are these versions being decreased?

File: Scripts/AppCenter/build-appcenter-ios-framework.sh (hunk 8 above)
Comment (jwallra) [reply Scripts/AppCenter/build-appcenter-ios-framework.sh]: I don't know if you copied from the script somewhere @jaelim-ms, but if you did then it's *`legacy`.  Also FAT should be capitalized

File: CHANGELOG.md (hunk 4 above)
Comment (ElektrojungeMS) [reply CHANGELOG.md]: Good point.

File: Config/iOS.xcconfig (hunk 5 above)
Comment (ElektrojungeMS) [reply Config/iOS.xcconfig]: yes, they are absolutely required.
Without specifying the versions individually, it will not work.

File: Scripts/AppCenterAnalytics/build-appcenteranalytics-tvos-framework.sh (hunk 6 above)
Comment (ElektrojungeMS) [reply Scripts/AppCenterAnalytics/build-appcenteranalytics-tvos-framework.sh]: yes

Comment (jwallra) [reply Scripts/AppCenterAnalytics/build-appcenteranalytics-tvos-framework.sh]: I would think so... It's just a comment, and wasn't the point of this to make a frankenbinary just like it's describing here?

Comment (jwallra) [reply Scripts/AppCenterAnalytics/build-appcenteranalytics-tvos-framework.sh]: Oh I see, it's for tvOS... is tvOS available for arm64e?

File: Scripts/AppCenter/build-appcenter-ios-framework.sh (hunk 9 above)
Comment (ElektrojungeMS) [reply Scripts/AppCenter/build-appcenter-ios-framework.sh]: I don't understand the question, the way above is how this works. I'm happy to optimize and spend more time on this. Can you please create a bug for this and we can discuss if we want to optimize?

File: CHANGELOG.md [hunk 12]
```diff
@@ -9,6 +9,10 @@
...
+
+* **[Fix]** Add preview support for arm64e CPU architectures by using PLCrashReporter 1.2.3-rc1.
```
[the remark this answers is not in the document]
Comment (ElektrojungeMS) [reply CHANGELOG.md]: Good idea.

File: README.md (hunk 10 above)
Comment (ElektrojungeMS) [reply README.md]: It's Mojave.

File: Sasquatch/Sasquatch/Base.lproj/LaunchScreen.storyboard (hunk 11 above)
Comment (ElektrojungeMS) [reply Sasquatch/Sasquatch/Base.lproj/LaunchScreen.storyboard]: We opened the storyboard in Xcode 8.3.3 which causes those changes as Xcode 8 does not know about `retina5_9`.

File: README.md (hunk 10 above)
Comment (jwallra) [reply README.md]: Right, so what about people on Mojave?  Are we just saying that they can't contribute until arm64e is out of preview?

**Comment by ElektrojungeMS** [comment]
@jwallra I cannot answer your questions for some reason.

1. Today `arm64e` is iOS only. My suspicion is that the next Apple TV will have an arm64e chip but that's for Apple to be decide ;)
2. `FAT` vs `fat`: `FAT` is the file format, a "fat file" doesn't have a capitalization.

File: Scripts/AppCenterAnalytics/build-appcenteranalytics-tvos-framework.sh (hunk 6 above)
Comment (ElektrojungeMS) [reply Scripts/AppCenterAnalytics/build-appcenteranalytics-tvos-framework.sh]: No, iOS only (see comment below).

File: Scripts/AppCenter/build-appcenter-ios-framework.sh (hunk 8 above)
Comment (achocron) [reply Scripts/AppCenter/build-appcenter-ios-framework.sh]: @jaelim-ms I think we should do general script optimizations in a bug since we have other things - like tons of duplicated code. We could do your suggestion there. Does that make sense?

File: README.md (hunk 10 above)
Comment (achocron) [reply README.md]: In my opinion, if someone is contributing, they don't need to build the SDK the same way we do at release time with CI. So I actually think we should tell people to use Xcode 10.1.

Comment (jwallra) [reply README.md]: That seems to be a good compromise.  We can also have instructions for testing out the arm64e preview for people who want to do it.

**Comment by jwallra** [comment]
> @jwallra I cannot answer your questions for some reason.
> 
> 1. Today `arm64e` is iOS only. My suspicion is that the next Apple TV will have an arm64e chip but that's for Apple to be decide ;)
> 2. `FAT` vs `fat`: `FAT` is the file format, a "fat file" doesn't have a capitalization.

@ElektrojungeMS Ah K, didn't know the distinction about FAT vs. fat.  Thanks :)

File: Scripts/AppCenter/build-appcenter-ios-framework.sh (hunk 9 above)
Comment (ElektrojungeMS) [reply Scripts/AppCenter/build-appcenter-ios-framework.sh]: Maybe do this with Zander's suggestion?

**Comment by ElektrojungeMS** [comment]
This is ready to merge!

File: Scripts/AppCenter/build-appcenter-ios-framework.sh (hunk 9 above)
Comment (jaelim-ms) [reply Scripts/AppCenter/build-appcenter-ios-framework.sh]: In other words, you are declaring an environment variable here for `DEVELOPER_DIR` but this isn't used anywhere. The line executes a commend and put result (something that is printed on console) to `DEVELOPER_DIR` variable.

File: Scripts/AppCenter/build-appcenter-ios-framework.sh (hunk 7 above)
Comment (jaelim-ms) [reply Scripts/AppCenter/build-appcenter-ios-framework.sh]: It doesn't work if you already have Xcode open and put an environment variable in Terminal after that. Once you reopen Xcode, it should be able to read the variable. This is just one time thing so we don't need to execute `source` to reload `.bash_profile`.
Just imagine when you install Android Studio without `JAVA_HOME`. You should close your IDE and open it again after setting the variable. I can explain this offline.

File: Scripts/AppCenter/build-appcenter-ios-framework.sh (hunk 8 above)
Comment (jaelim-ms) [reply Scripts/AppCenter/build-appcenter-ios-framework.sh]: What I meant was that the code between line 50 and 72 could be written like below.
```
if [ -z "$MS_ARM64E_XCODE_PATH" ] || [ ! -d "$MS_ARM64E_XCODE_PATH" ] ; then
  echo "Environment variable MS_ARM64E_XCODE_PATH not set or not a valid path."
  # Uses the Lipo tool to merge both binary files (i386/x86_64 + armv7/armv7s/arm64/arm64e) into one universal final product.
  echo "Use legacy Xcode and lipo -create the fat binary."
  lipo -create "${LIB_IPHONEOS_FINAL}" "${SIMULATOR_DIR}/lib${FMK_NAME}.a" -output 
  "${INSTALL_DIR}/${FMK_NAME}"
else
  echo "Building the arm64e slice."
  LIB_IPHONEOS_TEMP_DIR="${DEVICE_DIR}/temp"
  mkdir -p "${LIB_IPHONEOS_TEMP_DIR}"
  mv "${DEVICE_DIR}/lib${FMK_NAME}.a" "${LIB_IPHONEOS_TEMP_DIR}/lib${FMK_NAME}.a"

  # Build with the Xcode version that supports arm64e.
  env DEVELOPER_DIR="${MS_ARM64E_XCODE_PATH}" /usr/bin/xcodebuild ARCHS="arm64e" - project "${FMK_NAME}.xcodeproj" -configuration "Release" -target "${TGT_NAME}"

  # Lipo the binaries that were built from various Xcode versions.
  env DEVELOPER_DIR="${MS_ARM64E_XCODE_PATH}" /usr/bin/lipo -create 
  "${LIB_IPHONEOS_TEMP_DIR}/lib${FMK_NAME}.a" "${LIB_IPHONEOS_FINAL}" -output 
  "${LIB_IPHONEOS_FINAL}"

  # Uses the Lipo tool to merge both binary files (i386/x86_64 + armv7/armv7s/arm64/arm64e) into one universal final product.
  echo "Use arm64e Xcode and lipo -create the fat binary."
  env DEVELOPER_DIR="$MS_ARM64E_XCODE_PATH" /usr/bin/lipo -create 
  "${LIB_IPHONEOS_FINAL}" "${SIMULATOR_DIR}/lib${FMK_NAME}.a" -output 
  "${INSTALL_DIR}/${FMK_NAME}"
fi
```

File: Scripts/AppCenter/build-appcenter-ios-framework.sh (hunk 9 above)
Comment (jaelim-ms) [reply Scripts/AppCenter/build-appcenter-ios-framework.sh]: If you'd like it separately, it is fine.

File: Scripts/AppCenter/build-appcenter-ios-framework.sh (hunk 8 above)
Comment (achocron) [reply Scripts/AppCenter/build-appcenter-ios-framework.sh]: Made a [bug](https://msmobilecenter.visualstudio.com/Mobile-Center/_workitems/edit/53938)

File: Scripts/AppCenter/build-appcenter-ios-framework.sh (hunk 9 above)
Comment (achocron) [reply Scripts/AppCenter/build-appcenter-ios-framework.sh]: Made a [bug](https://msmobilecenter.visualstudio.com/Mobile-Center/_workitems/edit/53938)

File: Scripts/AppCenter/build-appcenter-ios-framework.sh (hunk 8 above)
Comment (ElektrojungeMS) [reply Scripts/AppCenter/build-appcenter-ios-framework.sh]: So we've actually cleaned this up a bit this morning. The bug that @achocron created is still valid because of duplicate code across the build scripts.

File: Scripts/AppCenter/build-appcenter-ios-framework.sh (hunk 9 above)
Comment (jaelim-ms) [reply Scripts/AppCenter/build-appcenter-ios-framework.sh]: Synced with @clpolet. `xcodebuild` and `lipo` read this variable so we need this one.
Please ignore my initial comment.

File: Scripts/AppCenter/build-appcenter-ios-framework.sh (hunk 7 above)
Comment (jaelim-ms) [reply Scripts/AppCenter/build-appcenter-ios-framework.sh]: I also synced with @clpolet. Then my suggestion would be using another file other than `.bash_profile`.
You can add a new file in the root of the repo and add the file to `.gitignore`. Then export MS_ARM64E_XCODE_PATH in the new file as you did in `~/.bash_profile` and `source the_new_file` in the script.
